awssql/internal_pool/internal_conn_pool.go

```diff
@@ -0,0 +1,196 @@
+/*
+  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+  Licensed under the Apache License, Version 2.0 (the "License").
+  You may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+package internal_pool
+
+import (
+	"context"
+	"database/sql/driver"
+	"log/slog"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-advanced-go-wrapper/awssql/error_util"
+)
+
+type internalPooledConn struct {
+	driver.Conn
+	pool       *InternalConnPool
+	createdAt  time.Time
+	lastUsedAt time.Time
+
+	mu       sync.Mutex
+	returned bool
+}
+
+func (pc *internalPooledConn) Close() error {
+	pc.mu.Lock()
+	if pc.returned {
+		pc.mu.Unlock()
+		return nil
+	}
+	pc.mu.Unlock()
```

Comment:
```suggestion
	defer pc.mu.Unlock()
```
What is this mutex protecting? You release it before changing the `pc.returned` value?

Reply:
That's a bug on my part. It should've been after. I'm going to move it after pc.returned = true instead of doing a defer as it would be more efficient than locking the whole function